File: javascript/qr_crypto.js
// qr_crypto.js

// Check if we're in a Node.js environment and use a polyfill if necessary
if (typeof window === 'undefined') {
    global.crypto = require('crypto').webcrypto;
}

class QRCrypto {
    static async generateQRKeys() {
        const keyPair = await crypto.subtle.generateKey(
            {
                name: "ECDSA",
                namedCurve: "P-256"
            },
            true,
            ["sign", "verify"]
        );

        const privateKey = await crypto.subtle.exportKey("pkcs8", keyPair.privateKey);
        const publicKey = await crypto.subtle.exportKey("spki", keyPair.publicKey);

        return {
            privateKey: this._arrayBufferToBase64(privateKey),
            publicKey: this._arrayBufferToBase64(publicKey)
        };
    }

    static async signQRTransaction(privateKeyBase64, transactionData) {
        const privateKey = await crypto.subtle.importKey(
            "pkcs8",
            this._base64ToArrayBuffer(privateKeyBase64),
            {
                name: "ECDSA",
                namedCurve: "P-256"
            },
            false,
            ["sign"]
        );

        const signature = await crypto.subtle.sign(
            {
                name: "ECDSA",
                hash: {name: "SHA-256"},
            },
            privateKey,
            new TextEncoder().encode(transactionData)
        );

        return this._arrayBufferToBase64(signature);
    }

    static async verifyQRSignature(publicKeyBase64, transactionData, signatureBase64) {
        const publicKey = await crypto.subtle.importKey(
            "spki",
            this._base64ToArrayBuffer(publicKeyBase64),
            {
                name: "ECDSA",
                namedCurve: "P-256"
            },
            false,
            ["verify"]
        );

        return await crypto.subtle.verify(
            {
                name: "ECDSA",
                hash: {name: "SHA-256"},
            },
            publicKey,
            this._base64ToArrayBuffer(signatureBase64),
            new TextEncoder().encode(transactionData)
        );
    }

    static _arrayBufferToBase64(buffer) {
        let binary = '';
        const bytes = new Uint8Array(buffer);
        for (let i = 0; i < bytes.byteLength; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return btoa(binary);
    }

    static _base64ToArrayBuffer(base64) {
        const binaryString = atob(base64);
        const bytes = new Uint8Array(binaryString.length);
        for (let i = 0; i < binaryString.length; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
        return bytes.buffer;
    }
}

// Export the class for Node.js environments
if (typeof module !== 'undefined' && module.exports) {
    module.exports = QRCrypto;
}

File: python/src/qr_crypto.py
from Crypto.PublicKey import ECC
from Crypto.Signature import DSS
from Crypto.Hash import SHA256
import binascii

def generate_qr_keys():
    key = ECC.generate(curve='P-256')
    private_key = key.export_key(format='PEM')
    public_key = key.public_key().export_key(format='PEM')
    return private_key, public_key

def sign_qr_transaction(private_key, transaction_data):
    key = ECC.import_key(private_key)
    h = SHA256.new(transaction_data.encode())
    signer = DSS.new(key, 'fips-186-3')
    signature = signer.sign(h)
    return binascii.hexlify(signature).decode()

def verify_qr_signature(public_key, transaction_data, signature):
    key = ECC.import_key(public_key)
    h = SHA256.new(transaction_data.encode())
    verifier = DSS.new(key, 'fips-186-3')
    try:
        verifier.verify(h, binascii.unhexlify(signature))
        return True
    except ValueError:
        return False

# Test the functions
if __name__ == "__main__":
    priv, pub = generate_qr_keys()
    print(f"Private Key: {priv}")
    print(f"Public key: {pub}")
    message = "Test transaction"
    signature = sign_qr_transaction(priv, message)
    print(f"Signature: {signature}")
    print(f"Verification: {verify_qr_signature(pub, message, signature)}")
